rfd/0212-jsonpath-interpolation.md

```diff
@@ -0,0 +1,647 @@
+version: v2
+metadata:
+  name: arbitrary-json-idp
+spec:
+  ...
+  claims_to_roles:
+    - claim: "roles"
+      value: "template"
+      roles: ["template"]
+```
+
+Lastly, the admin can create the template role and utilize the mapped traits.
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: template
+spec:
+  allow:
+    logins: '{{external.logins}}'
+    node_labels:
+      '*': '{{external.node_labels_*}}'
+      'env': '{{external.node_labels_env}}'
+    app_labels:
+      '*': '{{external.app_labels_*}}'
+      'env': '{{external.app_labels_env}}'
+```
+
+In the end, Alice's effective role will be:
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: template
+spec:
+  allow:
+    logins: ['alice']
+    node_labels:
+      '*': '*'
+    app_labels:
+      'env': 'staging'
+```
+
+##### Example: Distributed IdP
+
+Imagine a distributed IdP that aggregates claims for a user from multiple
+different provider sources, where each provider is associated with a different
+set of resources in Teleport.
+
+```json
+{
+  "aggregated_claims": {
+    "okta": {
+      "groups": ["teleport-access", "github-admin"],
+      "logins": "alice",
+      "env": ["staging", "dev"]
+    },
+    "auth0": {
+      "groups": "teleport-devops",
+      "logins": "devops",
+      "env": ["prod"]
+    },
+    "github": {
+      // no claims from github for this user.
+    }
+  }
+}
+```
+
+For this example, we will start with the OIDC connector `claims_to_roles` spec,
+which also supports the `jsonpath` function.
+
+```yaml
+kind: oidc 
+version: v2
+metadata:
+  name: distributed-idp
+spec:
+  ...
+  claims_to_roles:
+    - claim_expression: "jsonpath(external.aggregated_claims, "$.okta.groups")"
+      value: "teleport-access"
+      roles: ["okta-access"]
+      # evaluates to "teleport-devops", match.
+    - claim_expression: "jsonpath(external.aggregated_claims, "$.auth0.groups")"
+      value: "teleport-devops"
+      roles: ["auth0-devops"]
+      # evaluates to [], no match.
+    - claim_expression: "jsonpath(external.aggregated_claims, "$.github.groups")"
+      value: "*-admin"
+      roles: ["github-admin"]
+```
+
+In order to save the other claims as user traits, the admin can create a
+`login_rule`.
+
+The admin can also set other custom traits based on the structure of the claims.
+For example, a `teams` trait that aggregates the root property names of the
+aggregated claims (e.g. `okta`).
+
+```yaml
+kind: login_rule
+version: v1
+metadata:
+  name: distributed-idp
+spec:
+  priority: 0
+  traits_map:
+    okta_logins:
+      # evaluates to "alice"
+      - jsonpath(external.aggregated_claims, "$.okta.logins")
+    okta_env:
+      # evaluates to ["staging", "dev"]
+      - jsonpath(external.aggregated_claims, "$.okta.env")
+    auth0_logins:
+      # evaluates to "devops"
+      - jsonpath(external.aggregated_claims, "$.auth0.logins")
+    auth0_env:
+      # evaluates to ["prod"]
+      - jsonpath(external.aggregated_claims, "$.auth0.env")
```

Comment:
when I see this i think why not aggregate the logins and envs here rather than using 2 roles that consume the different ones?
```suggestion
    logins:
      # evaluates to "alice"
      - jsonpath(external.aggregated_claims, "$.okta.logins")
      # evaluates to "devops"
      - jsonpath(external.aggregated_claims, "$.auth0.logins")
    envs:
      # evaluates to ["staging", "dev"]
      - jsonpath(external.aggregated_claims, "$.okta.env")
      # evaluates to ["prod"]
      - jsonpath(external.aggregated_claims, "$.auth0.env")
```

Reply:
yeah i see, i missed that aspect